src/Features/CSharp/Portable/CodeRefactorings/UseRecursivePatterns/UseRecursivePatternsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,545 @@
+                var list = new SyntaxNodeOrTokenList(leftSubpatterns.Concat(rightSubpatterns));
+                return left.WithSubpatterns(SeparatedList<SubpatternSyntax>(list));
+            }
+        }
+
+        private static PatternSyntax CreatePattern(ExpressionSyntax receiver, ExpressionOrPatternSyntax target, bool flipped)
+        {
+            return target switch
+            {
+                // A type or pattern come from an `is` expression on either side of `&&`
+                PatternSyntax pattern => pattern,
+                TypeSyntax type => TypePattern(type),
+                // Otherwise this is a constant. Depending on the original receiver, we create an appropriate pattern.
+                ExpressionSyntax constant => CreatePattern(receiver, constant, flipped),
+                var v => throw ExceptionUtilities.UnexpectedValue(v),
+            };
+
+            static PatternSyntax CreatePattern(ExpressionSyntax receiver, ExpressionSyntax target, bool flipped)
+            {
+                return receiver.Parent switch
+                {
+                    BinaryExpressionSyntax(EqualsExpression) => ConstantPattern(target),
+                    BinaryExpressionSyntax(NotEqualsExpression) => UnaryPattern(ConstantPattern(target)),
+                    BinaryExpressionSyntax(GreaterThanExpression or
+                                           GreaterThanOrEqualExpression or
+                                           LessThanOrEqualExpression or
+                                           LessThanExpression) e
+                        => RelationalPattern(flipped ? Flip(e.OperatorToken) : e.OperatorToken, target),
+                    var v => throw ExceptionUtilities.UnexpectedValue(v),
+                };
+            }
+
+            static SyntaxToken Flip(SyntaxToken token)
+            {
+                var kind = token.Kind() switch
+                {
+                    LessThanToken => GreaterThanToken,
+                    LessThanEqualsToken => GreaterThanEqualsToken,
+                    GreaterThanEqualsToken => LessThanEqualsToken,
+                    GreaterThanToken => LessThanToken,
+                    var v => throw ExceptionUtilities.UnexpectedValue(v)
+                };
+                return Token(token.LeadingTrivia, kind, token.TrailingTrivia);
+            }
+        }
+
+        private static (PatternSyntax ContainingPattern, ImmutableArray<IdentifierNameSyntax> NamesOpt)? TryFindVariableDesignation(
+            PatternSyntax leftPattern,
+            ExpressionSyntax rightReceiver,
+            SemanticModel model)
+        {
+            if (GetInnermostReceiver(rightReceiver, out var namesOpt, model) is not IdentifierNameSyntax identifierName)
+                return null;
+
+            var designation = leftPattern.DescendantNodes()
+                .OfType<SingleVariableDesignationSyntax>()
+                .Where(d => AreEquivalent(d.Identifier, identifierName.Identifier))
```

Comment:
instead of AreEquivalent, i think you just want `d.Identifier.ValueText == identifierNAme.Identifier.ValueText`